File: mkdocs/snippets/devbook/transactions/cross_chain_swap.py
import hashlib
import json
import os
import secrets
import time
import urllib.request

from symbolchain.CryptoTypes import Hash256, PrivateKey
from symbolchain.facade.SymbolFacade import SymbolFacade
from symbolchain.sc import Amount
from symbolchain.symbol.IdGenerator import generate_mosaic_alias_id
from symbolchain.symbol.Network import NetworkTimestamp
from web3 import Web3

SYMBOL_NODE_URL = os.getenv(
	'SYMBOL_NODE_URL', 'https://reference.symboltest.net:3001')
print(f'Using Symbol node {SYMBOL_NODE_URL}')

ETH_RPC_URL = os.getenv('ETH_RPC_URL',
	'https://ethereum-sepolia-rpc.publicnode.com')
print(f'Using Ethereum RPC {ETH_RPC_URL}')

# Ethereum HTLC contract on Sepolia
HTLC_ADDRESS = '0xd58e030bd21c7788897aE5Ea845DaBA936e91D2B'
HTLC_ABI = [
	{
		'name': 'newContract',
		'type': 'function',
		'stateMutability': 'payable',
		'inputs': [
			{'name': '_receiver', 'type': 'address'},
			{'name': '_hashlock', 'type': 'bytes32'},
			{'name': '_timelock', 'type': 'uint256'}
		],
		'outputs': [{'name': 'contractId', 'type': 'bytes32'}]
	},
	{
		'name': 'withdraw',
		'type': 'function',
		'stateMutability': 'nonpayable',
		'inputs': [
			{'name': '_contractId', 'type': 'bytes32'},
			{'name': '_preimage', 'type': 'bytes'}
		],
		'outputs': [{'name': '', 'type': 'bool'}]
	},
	{
		'name': 'getContract',
		'type': 'function',
		'stateMutability': 'view',
		'inputs': [
			{'name': '_contractId', 'type': 'bytes32'}
		],
		'outputs': [
			{'name': 'sender', 'type': 'address'},
			{'name': 'receiver', 'type': 'address'},
			{'name': 'amount', 'type': 'uint256'},
			{'name': 'hashlock', 'type': 'bytes32'},
			{'name': 'timelock', 'type': 'uint256'},
			{'name': 'withdrawn', 'type': 'bool'},
			{'name': 'refunded', 'type': 'bool'},
			{'name': 'preimage', 'type': 'bytes'}
		]
	},
	{
		'name': 'LogHTLCNew',
		'type': 'event',
		'inputs': [
			{'name': 'contractId', 'type': 'bytes32', 'indexed': True},
			{'name': 'sender', 'type': 'address', 'indexed': True},
			{'name': 'receiver', 'type': 'address', 'indexed': True},
			{'name': 'amount', 'type': 'uint256', 'indexed': False},
			{'name': 'hashlock', 'type': 'bytes32', 'indexed': False},
			{'name': 'timelock', 'type': 'uint256', 'indexed': False}
		]
	}
]


# Helper function to fetch current Symbol network time
def get_network_time():
	time_path = '/node/time'
	print(f'Fetching current network time from {time_path}')
	with urllib.request.urlopen(
		f'{SYMBOL_NODE_URL}{time_path}'
	) as response:
		response_json = json.loads(response.read().decode())
		timestamp = NetworkTimestamp(int(
			response_json['communicationTimestamps'][
				'receiveTimestamp']))
		print(f'  Network time: {timestamp.timestamp}'
			' ms since nemesis')
		return timestamp


# Helper function to fetch recommended Symbol fee multiplier
def get_fee_multiplier():
	fee_path = '/network/fees/transaction'
	print(f'Fetching recommended fees from {fee_path}')
	with urllib.request.urlopen(
		f'{SYMBOL_NODE_URL}{fee_path}'
	) as response:
		response_json = json.loads(response.read().decode())
		median_multiplier = response_json['medianFeeMultiplier']
		minimum_multiplier = response_json['minFeeMultiplier']
		fee_multiplier = max(median_multiplier, minimum_multiplier)
		print(f'  Fee multiplier: {fee_multiplier}')
		return fee_multiplier


# Helper function to announce a Symbol transaction
def announce_transaction(payload, endpoint, label):
	print(f'Announcing {label} to {endpoint}')
	request = urllib.request.Request(
		f'{SYMBOL_NODE_URL}{endpoint}',
		data=payload.encode(),
		headers={'Content-Type': 'application/json'},
		method='PUT'
	)
	with urllib.request.urlopen(request) as response:
		print(f'  Response: {response.read().decode()}')


# Helper function to wait for Symbol transaction status
def wait_for_status(hash_value, expected_status, label):
	print(f'Waiting for {label} to reach {expected_status} status...')
	attempts = 0
	max_attempts = 60

	while attempts < max_attempts:
		try:
			url = f'{SYMBOL_NODE_URL}/transactionStatus/{hash_value}'
			with urllib.request.urlopen(url) as response:
				status = json.loads(response.read().decode())
				print(f'  Transaction status: {status["group"]}')

				if status['group'] == 'failed':
					raise RuntimeError(f'{label} failed: {status["code"]}')

				if status['group'] == expected_status:
					print(f'{label} {expected_status}'
						f' in {attempts} seconds')
					return

		except urllib.error.HTTPError as err:
			if err.code != 404:
				raise
			print('  Transaction status: not yet available')

		attempts += 1
		time.sleep(1)

	raise TimeoutError(
		f'{label} not {expected_status} after {max_attempts} attempts')


# Poll Symbol for a confirmed secret proof transaction matching
# a hashlock.
def wait_for_secret_proof(signer_address, hlock):
	hashlock_hex = hlock.hex().upper()
	url = (f'{SYMBOL_NODE_URL}/transactions/confirmed'
		f'?address={signer_address}&type=16978&order=desc')
	print(f'Polling {url}')
	print(f'  Looking for secret: {hashlock_hex}')

	attempts = 0
	max_attempts = 60
	while attempts < max_attempts:
		with urllib.request.urlopen(url) as response:
			data = json.loads(response.read().decode())
		for trans in data.get('data', []):
			found_secret = trans['transaction'].get('secret', '')
			if found_secret.upper() == hashlock_hex:
				print(f'  Found proof transaction after {attempts}s')
				return bytes.fromhex(trans['transaction']['proof'])
		attempts += 1
		time.sleep(1)

	raise TimeoutError(
		f'Secret proof not found after {max_attempts} attempts')


# Symbol accounts [>step-1]
facade = SymbolFacade('testnet')

ALICE_XYM_PRIVATE_KEY = os.getenv('ALICE_XYM_PRIVATE_KEY',
	'0000000000000000000000000000000000000000000000000000000000000000')
alice_xym_key_pair = SymbolFacade.KeyPair(
	PrivateKey(ALICE_XYM_PRIVATE_KEY))
alice_xym_address = facade.network.public_key_to_address(
	alice_xym_key_pair.public_key)
print(f'Alice Symbol address: {alice_xym_address}')

BOB_XYM_PRIVATE_KEY = os.getenv('BOB_XYM_PRIVATE_KEY',
	'1111111111111111111111111111111111111111111111111111111111111111')
bob_xym_key_pair = SymbolFacade.KeyPair(PrivateKey(BOB_XYM_PRIVATE_KEY))
bob_xym_address = facade.network.public_key_to_address(
	bob_xym_key_pair.public_key)
print(f'Bob Symbol address: {bob_xym_address}')

# Ethereum accounts
w3 = Web3(Web3.HTTPProvider(ETH_RPC_URL))

ALICE_ETH_PRIVATE_KEY = os.getenv('ALICE_ETH_PRIVATE_KEY',
	'0xa73276699ba72dc7b5c9d08deaf8cd88eec33c866341b120304432b89586d45d')
alice_eth_account = w3.eth.account.from_key(ALICE_ETH_PRIVATE_KEY)
print(f'Alice ETH address: {alice_eth_account.address}')

BOB_ETH_PRIVATE_KEY = os.getenv('BOB_ETH_PRIVATE_KEY',
	'0x8e85561005f27d926af79a7ce3e76e75108a09ff2ab78bf65b5578d2e5d509bf')
bob_eth_account = w3.eth.account.from_key(BOB_ETH_PRIVATE_KEY)
print(f'Bob ETH address: {bob_eth_account.address}')
# [<step-1]
try:
	# --- Alice: Generate proof and hashlock ---
	print('\n--- Alice: Generate proof and hashlock ---')  # [>step-2]

	proof = secrets.token_bytes(32)
	print(f'Proof (hex): {proof.hex()}')

	first_hash = hashlib.sha256(proof).digest()
	secret = hashlib.sha256(first_hash).digest()
	print(f'Secret (double SHA-256): {secret.hex()}')
	# [<step-2]
	# --- Step 1. Alice: Lock ETH on Ethereum ---
	print('\n--- Step 1. Alice: Lock ETH on Ethereum ---')  # [>step-3]

	htlc = w3.eth.contract(address=HTLC_ADDRESS, abi=HTLC_ABI)
	timelock = int(time.time()) + 72 * 60 * 60
	print(f'Ethereum timelock (Unix): {timelock}')

	lock_call = htlc.functions.newContract(
		bob_eth_account.address, secret, timelock)
	lock_tx = lock_call.build_transaction({
		'from': alice_eth_account.address,
		'value': w3.to_wei(0.01, 'ether'),
		'nonce': w3.eth.get_transaction_count(alice_eth_account.address)
	})
	signed_lock_tx = alice_eth_account.sign_transaction(lock_tx)
	lock_tx_hash = w3.eth.send_raw_transaction(
		signed_lock_tx.raw_transaction)
	print(f'Lock TX hash: {lock_tx_hash.hex()}')

	lock_receipt = w3.eth.wait_for_transaction_receipt(lock_tx_hash)
	print(f'Lock confirmed in block {lock_receipt.blockNumber}')

	# Extract the contractId from the LogHTLCNew event
	contract_id = lock_receipt.logs[0].topics[1]
	print(f'HTLC contract ID: {contract_id.hex()}')
	# [<step-3]
	# --- Step 2. Bob: Create secret lock on Symbol ---
	print('\n--- Step 2. Bob: Create secret lock on Symbol ---')  # [>step-4]

	# Bob queries the Ethereum contract to get the hashlock
	contract_info = htlc.functions.getContract(contract_id).call()
	hashlock = contract_info[3]  # hashlock field
	print(f'Hashlock from chain: {hashlock.hex()}')

	lock_duration = 5760  # ~48h at 30s blocks
	print(f'Lock duration: {lock_duration} blocks')

	secret_lock_transaction = facade.transaction_factory.create({
		'type': 'secret_lock_transaction_v1',
		'signer_public_key': bob_xym_key_pair.public_key,
		'deadline': get_network_time().add_hours(2).timestamp,
		'recipient_address': alice_xym_address,
		'mosaic': {
			'mosaic_id': generate_mosaic_alias_id('symbol.xym'),
			'amount': 1_000000  # 1 XYM
		},
		'duration': lock_duration,
		'secret': Hash256(hashlock),
		'hash_algorithm': 'hash_256'
	})
	secret_lock_transaction.fee = Amount(
		get_fee_multiplier() * secret_lock_transaction.size)

	# Sign and announce
	lock_signature = facade.sign_transaction(
		bob_xym_key_pair, secret_lock_transaction)
	lock_payload = facade.transaction_factory.attach_signature(
		secret_lock_transaction, lock_signature)

	print('Built secret lock transaction:')
	print(json.dumps(secret_lock_transaction.to_json(), indent=2))

	lock_hash = facade.hash_transaction(secret_lock_transaction)
	print(f'Secret lock transaction hash: {lock_hash}')
	announce_transaction(lock_payload, '/transactions', 'secret lock')
	wait_for_status(lock_hash, 'confirmed', 'Secret lock')
	# [<step-4]
	# --- Step 3. Alice: Claim XYM on Symbol ---
	print('\n--- Step 3. Alice: Claim XYM on Symbol ---')  # [>step-5]

	secret_proof_transaction = facade.transaction_factory.create({
		'type': 'secret_proof_transaction_v1',
		'signer_public_key': alice_xym_key_pair.public_key,
		'deadline': get_network_time().add_hours(2).timestamp,
		'recipient_address': alice_xym_address,
		'secret': Hash256(hashlock),
		'hash_algorithm': 'hash_256',
		'proof': proof
	})
	secret_proof_transaction.fee = Amount(
		get_fee_multiplier() * secret_proof_transaction.size)

	# Sign and announce
	proof_signature = facade.sign_transaction(
		alice_xym_key_pair, secret_proof_transaction)
	proof_payload = facade.transaction_factory.attach_signature(
		secret_proof_transaction, proof_signature)

	print('Built secret proof transaction:')
	print(json.dumps(secret_proof_transaction.to_json(), indent=2))

	proof_hash = facade.hash_transaction(secret_proof_transaction)
	print(f'Secret proof transaction hash: {proof_hash}')
	announce_transaction(proof_payload, '/transactions', 'secret proof')
	wait_for_status(proof_hash, 'confirmed', 'Secret proof')
	# [<step-5]
	# --- Step 4. Bob: Withdraw ETH on Ethereum ---
	print('\n--- Step 4. Bob: Withdraw ETH on Ethereum ---')  # [>step-6]

	# Bob waits for Alice to reveal the proof on Symbol.
	revealed_proof = wait_for_secret_proof(alice_xym_address, hashlock)
	print(f'Proof from chain: {revealed_proof.hex()}')

	withdraw_call = htlc.functions.withdraw(contract_id, revealed_proof)
	withdraw_tx = withdraw_call.build_transaction({
		'from': bob_eth_account.address,
		'nonce': w3.eth.get_transaction_count(bob_eth_account.address)
	})
	signed_withdraw_tx = bob_eth_account.sign_transaction(withdraw_tx)
	withdraw_tx_hash = w3.eth.send_raw_transaction(
		signed_withdraw_tx.raw_transaction)
	print(f'Withdraw TX hash: {withdraw_tx_hash.hex()}')

	withdraw_receipt = w3.eth.wait_for_transaction_receipt(
		withdraw_tx_hash)
	print(f'Withdraw confirmed in block {withdraw_receipt.blockNumber}')
	# [<step-6]
	print('\n--- Cross-chain swap complete ---')

except urllib.error.URLError as e:
	print(e.reason)
except Exception as e:
	print(e)

File: mkdocs/snippets/devbook/transactions/cross_chain_swap.java
//JAVA 21+
//DEPS org.web3j:core:4.12.3
//DEPS org.symbol:symbol-sdk:3.3.1

import java.io.IOException;
import java.math.BigInteger;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Arrays;
import java.util.HexFormat;
import java.util.List;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.web3j.abi.FunctionEncoder;
import org.web3j.abi.FunctionReturnDecoder;
import org.web3j.abi.TypeReference;
import org.web3j.abi.datatypes.Bool;
import org.web3j.abi.datatypes.DynamicBytes;
import org.web3j.abi.datatypes.Function;
import org.web3j.abi.datatypes.generated.Bytes32;
import org.web3j.abi.datatypes.generated.Uint256;
import org.web3j.crypto.Credentials;
import org.web3j.protocol.Web3j;
import org.web3j.protocol.core.DefaultBlockParameterName;
import org.web3j.protocol.core.methods.response.EthCall;
import org.web3j.protocol.core.methods.response.TransactionReceipt;
import org.web3j.protocol.http.HttpService;
import org.web3j.tx.RawTransactionManager;
import org.web3j.tx.response.PollingTransactionReceiptProcessor;
import org.web3j.utils.Convert;
import org.web3j.utils.Numeric;

import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.facade.SymbolFacade;
import org.symbol.sdk.symbol.Address;
import org.symbol.sdk.symbol.IdGenerator;
import org.symbol.sdk.symbol.KeyPair;
import org.symbol.sdk.symbol.SymbolTransactionFactory;
import org.symbol.sdk.symbol.descriptors.*;
import org.symbol.sdk.symbol.models.*;
import org.symbol.sdk.utils.Converter;

final class CrossChainSwap {
	private static final ObjectMapper JSON_MAPPER = new ObjectMapper();

	private static final HttpClient HTTP_CLIENT =
		HttpClient.newHttpClient();

	private static final SecureRandom RANDOM = new SecureRandom();

	private final String symbolNodeUrl = System.getenv().getOrDefault(
		"SYMBOL_NODE_URL", "https://reference.symboltest.net:3001");

	private final String ethRpcUrl = System.getenv().getOrDefault(
		"ETH_RPC_URL", "https://ethereum-sepolia-rpc.publicnode.com");

	// Ethereum HTLC contract on Sepolia
	private static final String HTLC_ADDRESS =
		"0xd58e030bd21c7788897aE5Ea845DaBA936e91D2B";

	private final SymbolFacade facade = new SymbolFacade("testnet");

	private Web3j ethProvider;

	private Credentials aliceEthWallet;

	private Credentials bobEthWallet;

	private Address aliceXymAddress;

	private Address bobXymAddress;

	private KeyPair aliceXymKeyPair;

	private KeyPair bobXymKeyPair;

	public static void main(final String[] args) {
		try {
			new CrossChainSwap().run();
		} catch (final Exception ex) {
			System.out.println(null == ex.getMessage()
				? ex.toString()
				: ex.getMessage());
		}
	}

	private void run()
		throws Exception {
		System.out.printf("Using Symbol node %s%n", symbolNodeUrl);
		System.out.printf("Using Ethereum RPC %s%n", ethRpcUrl);

		// Symbol accounts [>step-1]
		// Alice (creates the ETH lock, claims XYM on Symbol)
		final String aliceXymPrivateKey = System.getenv().getOrDefault(
			"ALICE_XYM_PRIVATE_KEY", "0".repeat(64));
		aliceXymKeyPair = new KeyPair(
			new CryptoTypes.PrivateKey(aliceXymPrivateKey));
		aliceXymAddress = facade.network.publicKeyToAddress(
			aliceXymKeyPair.getPublicKey());
		System.out.printf("Alice Symbol address: %s%n", aliceXymAddress);

		// Bob (creates the XYM lock, claims ETH on Ethereum)
		final String bobXymPrivateKey = System.getenv().getOrDefault(
			"BOB_XYM_PRIVATE_KEY", "1".repeat(64));
		bobXymKeyPair = new KeyPair(
			new CryptoTypes.PrivateKey(bobXymPrivateKey));
		bobXymAddress = facade.network.publicKeyToAddress(
			bobXymKeyPair.getPublicKey());
		System.out.printf("Bob Symbol address: %s%n", bobXymAddress);

		// Ethereum accounts
		ethProvider = Web3j.build(new HttpService(ethRpcUrl));

		final String aliceEthPrivateKey = System.getenv().getOrDefault(
			"ALICE_ETH_PRIVATE_KEY",
			"a73276699ba72dc7b5c9d08deaf8cd88eec33c866341b12030443"
				+ "2b89586d45d");
		aliceEthWallet = Credentials.create(aliceEthPrivateKey);
		System.out.printf("Alice ETH address: %s%n",
			aliceEthWallet.getAddress());

		final String bobEthPrivateKey = System.getenv().getOrDefault(
			"BOB_ETH_PRIVATE_KEY",
			"8e85561005f27d926af79a7ce3e76e75108a09ff2ab78bf65b557"
				+ "8d2e5d509bf");
		bobEthWallet = Credentials.create(bobEthPrivateKey);
		System.out.printf("Bob ETH address: %s%n",
			bobEthWallet.getAddress());
		// [<step-1]

		// --- Alice: Generate proof and hashlock ---
		System.out.println( // [>step-2]
			"\n--- Alice: Generate proof and hashlock ---");

		final byte[] proof = new byte[32];
		RANDOM.nextBytes(proof);
		System.out.printf("Proof (hex): %s%n",
			HexFormat.of().formatHex(proof));

		final byte[] firstHash = sha256(proof);
		final byte[] secret = sha256(firstHash);
		System.out.printf("Secret (double SHA-256): %s%n",
			HexFormat.of().formatHex(secret));
		// [<step-2]
		// --- Step 1. Alice: Lock ETH on Ethereum ---
		System.out.println( // [>step-3]
			"\n--- Step 1. Alice: Lock ETH on Ethereum ---");

		final long timelock =
			System.currentTimeMillis() / 1000 + (72 * 60 * 60);
		System.out.printf("Ethereum timelock (Unix): %d%n", timelock);

		final String lockHash = sendNewContract(
			bobEthWallet.getAddress(),
			secret,
			BigInteger.valueOf(timelock));
		System.out.printf("Lock TX hash: %s%n", lockHash);

		final TransactionReceipt lockReceipt = waitForEthereumReceipt(
			lockHash);
		System.out.printf("Lock confirmed in block %s%n",
			lockReceipt.getBlockNumber());

		// Extract the contractId from the LogHTLCNew event
		final String contractId = lockReceipt.getLogs().get(0)
			.getTopics().get(1);
		System.out.printf("HTLC contract ID: %s%n", contractId);
		// [<step-3]
		// --- Step 2. Bob: Create secret lock on Symbol ---
		System.out.println( // [>step-4]
			"\n--- Step 2. Bob: Create secret lock on Symbol ---");

		// Bob queries the Ethereum contract to get the hashlock
		final String hashlock = getContractHashlock(contractId);
		System.out.printf("Hashlock from chain: %s%n", hashlock);

		final long lockDuration = 5760; // ~48h at 30s blocks
		System.out.printf("Lock duration: %d blocks%n", lockDuration);

		final Transaction secretLockTransaction =
			facade.createTransactionFromTypedDescriptor(
				new SecretLockTransactionV1Descriptor(
					aliceXymAddress,
					new CryptoTypes.Hash256(hashlock),
					new UnresolvedMosaicDescriptor(
						new UnresolvedMosaicId(
							IdGenerator.generateMosaicAliasId(
								"symbol.xym")),
						new Amount(1_000_000)), // 1 XYM
					new BlockDuration(lockDuration),
					LockHashAlgorithm.HASH_256),
				bobXymKeyPair.getPublicKey(),
				getFeeMultiplier(),
				2 * 60 * 60);

		// Sign and announce
		final CryptoTypes.Signature lockSignature =
			facade.signTransaction(bobXymKeyPair, secretLockTransaction);
		final String lockPayload = SymbolTransactionFactory
			.attachSignature(secretLockTransaction, lockSignature);

		System.out.println("Built secret lock transaction:");
		System.out.println(JSON_MAPPER.writerWithDefaultPrettyPrinter()
			.writeValueAsString(secretLockTransaction.toJson()));

		final String symbolLockHash = facade.hashTransaction(
			secretLockTransaction).toString();
		System.out.printf("Secret lock transaction hash: %s%n",
			symbolLockHash);
		announceTransaction(lockPayload, "/transactions", "secret lock");
		waitForStatus(symbolLockHash, "confirmed", "Secret lock");
		// [<step-4]
		// --- Step 3. Alice: Claim XYM on Symbol ---
		System.out.println( // [>step-5]
			"\n--- Step 3. Alice: Claim XYM on Symbol ---");

		final Transaction secretProofTransaction =
			facade.createTransactionFromTypedDescriptor(
				new SecretProofTransactionV1Descriptor(
					aliceXymAddress,
					new CryptoTypes.Hash256(hashlock),
					LockHashAlgorithm.HASH_256,
					proof),
				aliceXymKeyPair.getPublicKey(),
				getFeeMultiplier(),
				2 * 60 * 60);

		// Sign and announce
		final CryptoTypes.Signature proofSignature =
			facade.signTransaction(
				aliceXymKeyPair, secretProofTransaction);
		final String proofPayload = SymbolTransactionFactory
			.attachSignature(secretProofTransaction, proofSignature);

		System.out.println("Built secret proof transaction:");
		System.out.println(JSON_MAPPER.writerWithDefaultPrettyPrinter()
			.writeValueAsString(secretProofTransaction.toJson()));

		final String proofHash = facade.hashTransaction(
			secretProofTransaction).toString();
		System.out.printf("Secret proof transaction hash: %s%n",
			proofHash);
		announceTransaction(proofPayload, "/transactions", "secret proof");
		waitForStatus(proofHash, "confirmed", "Secret proof");
		// [<step-5]
		// --- Step 4. Bob: Withdraw ETH on Ethereum ---
		System.out.println( // [>step-6]
			"\n--- Step 4. Bob: Withdraw ETH on Ethereum ---");

		// Bob waits for Alice to reveal the proof on Symbol.
		final byte[] revealedProof = waitForSecretProof(
			aliceXymAddress.toString(), hashlock);
		System.out.printf("Proof from chain: %s%n",
			HexFormat.of().formatHex(revealedProof));

		final String withdrawHash = sendWithdraw(
			contractId, revealedProof);
		System.out.printf("Withdraw TX hash: %s%n", withdrawHash);

		final TransactionReceipt withdrawReceipt = waitForEthereumReceipt(
			withdrawHash);
		System.out.printf("Withdraw confirmed in block %s%n",
			withdrawReceipt.getBlockNumber());
		// [<step-6]
		System.out.println("\n--- Cross-chain swap complete ---");
	}

	// Helper function to fetch recommended Symbol fee multiplier
	private long getFeeMultiplier()
		throws IOException, InterruptedException {
		final String feePath = "/network/fees/transaction";
		System.out.printf("Fetching recommended fees from %s%n", feePath);
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(symbolNodeUrl + feePath)).GET().build();
		final HttpResponse<String> response = HTTP_CLIENT.send(
			request, BodyHandlers.ofString());
		final JsonNode feeJSON = JSON_MAPPER.readTree(response.body());
		final long medianMultiplier =
			feeJSON.get("medianFeeMultiplier").asLong();
		final long minimumMultiplier =
			feeJSON.get("minFeeMultiplier").asLong();
		final long feeMultiplier = Math.max(
			medianMultiplier, minimumMultiplier);
		System.out.printf("  Fee multiplier: %d%n", feeMultiplier);
		return feeMultiplier;
	}

	// Helper function to announce a Symbol transaction
	private void announceTransaction(
		final String payload,
		final String endpoint,
		final String label
	) throws IOException, InterruptedException {
		System.out.printf("Announcing %s to %s%n", label, endpoint);
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(symbolNodeUrl + endpoint))
			.header("Content-Type", "application/json")
			.PUT(HttpRequest.BodyPublishers.ofString(payload))
			.build();
		final HttpResponse<String> response = HTTP_CLIENT.send(
			request, BodyHandlers.ofString());
		System.out.printf("  Response: %s%n", response.body());
	}

	// Helper function to wait for Symbol transaction status
	private void waitForStatus(
		final String hash,
		final String expectedStatus,
		final String label
	) throws IOException, InterruptedException {
		System.out.printf(
			"Waiting for %s to reach %s status...%n",
			label, expectedStatus);
		int attempts = 0;
		final int maxAttempts = 60;

		while (attempts < maxAttempts) {
			final String url = symbolNodeUrl + "/transactionStatus/"
				+ hash;
			final HttpRequest request = HttpRequest.newBuilder(
				URI.create(url)).GET().build();
			final HttpResponse<String> response = HTTP_CLIENT.send(
				request, BodyHandlers.ofString());

			if (response.statusCode() / 100 != 2) {
				if (404 == response.statusCode()) {
					System.out.println(
						"  Transaction status: not yet available");
				} else
					throw new IOException(
						"HTTP " + response.statusCode());
			} else {
				final JsonNode status = JSON_MAPPER.readTree(
					response.body());
				System.out.printf("  Transaction status: %s%n",
					status.get("group").asText());

				if ("failed".equals(status.get("group").asText()))
					throw new IOException(String.format("%s failed: %s",
						label, status.get("code").asText()));

				if (status.get("group").asText().equals(expectedStatus)) {
					System.out.printf("%s %s in %d seconds%n",
						label, expectedStatus, attempts);
					return;
				}
			}

			++attempts;
			Thread.sleep(1000);
		}

		throw new IOException(String.format(
			"%s not %s after %d attempts",
			label, expectedStatus, maxAttempts));
	}

	// Poll Symbol for a confirmed secret proof transaction matching
	// a hashlock.
	private byte[] waitForSecretProof(
		final String signerAddress,
		final String hashlock
	) throws IOException, InterruptedException {
		final String hashlockHex = hashlock.toUpperCase();
		final String url = symbolNodeUrl + "/transactions/confirmed"
			+ "?address=" + signerAddress + "&type=16978&order=desc";
		System.out.printf("Polling %s%n", url);
		System.out.printf("  Looking for secret: %s%n", hashlockHex);

		int attempts = 0;
		final int maxAttempts = 60;
		while (attempts < maxAttempts) {
			final HttpRequest request = HttpRequest.newBuilder(
				URI.create(url)).GET().build();
			final HttpResponse<String> response = HTTP_CLIENT.send(
				request, BodyHandlers.ofString());
			final JsonNode data = JSON_MAPPER.readTree(response.body());
			for (final JsonNode tx : data.get("data")) {
				final String secret = tx.get("transaction")
					.get("secret").asText("").toUpperCase();
				if (secret.equals(hashlockHex)) {
					System.out.printf(
						"  Found proof transaction after %ds%n",
						attempts);
					return Converter.hexToUint8(tx.get("transaction")
						.get("proof").asText());
				}
			}
			++attempts;
			Thread.sleep(1000);
		}

		throw new IOException(String.format(
			"Secret proof not found after %d attempts", maxAttempts));
	}

	private String sendNewContract(
		final String receiver,
		final byte[] hashlock,
		final BigInteger timelock
	) throws Exception {
		final Function function = new Function(
			"newContract",
			List.of(
				new org.web3j.abi.datatypes.Address(receiver),
				new Bytes32(hashlock),
				new Uint256(timelock)),
			List.of(new TypeReference<Bytes32>() {})
		);
		final String data = FunctionEncoder.encode(function);
		final RawTransactionManager manager = new RawTransactionManager(
			ethProvider, aliceEthWallet);
		return manager.sendTransaction(
			BigInteger.valueOf(20_000_000_000L),
			BigInteger.valueOf(300_000),
			HTLC_ADDRESS,
			data,
			Convert.toWei("0.01", Convert.Unit.ETHER).toBigInteger()
		).getTransactionHash();
	}

	private String getContractHashlock(
		final String contractId
	) throws IOException {
		final Function function = new Function(
			"getContract",
			List.of(new Bytes32(Numeric.hexStringToByteArray(contractId))),
			Arrays.asList(
				new TypeReference<org.web3j.abi.datatypes.Address>() {},
				new TypeReference<org.web3j.abi.datatypes.Address>() {},
				new TypeReference<Uint256>() {},
				new TypeReference<Bytes32>() {},
				new TypeReference<Uint256>() {},
				new TypeReference<Bool>() {},
				new TypeReference<Bool>() {},
				new TypeReference<DynamicBytes>() {})
		);
		final String data = FunctionEncoder.encode(function);
		final EthCall response = ethProvider.ethCall(
			org.web3j.protocol.core.methods.request.Transaction
				.createEthCallTransaction(
					bobEthWallet.getAddress(), HTLC_ADDRESS, data),
			DefaultBlockParameterName.LATEST
		).send();
		return Numeric.toHexStringNoPrefix((byte[]) FunctionReturnDecoder
			.decode(response.getValue(), function.getOutputParameters())
			.get(3).getValue());
	}

	private String sendWithdraw(
		final String contractId,
		final byte[] revealedProof
	) throws Exception {
		final Function function = new Function(
			"withdraw",
			List.of(
				new Bytes32(Numeric.hexStringToByteArray(contractId)),
				new DynamicBytes(revealedProof)),
			List.of(new TypeReference<Bool>() {})
		);
		final String data = FunctionEncoder.encode(function);
		final RawTransactionManager manager = new RawTransactionManager(
			ethProvider, bobEthWallet);
		return manager.sendTransaction(
			BigInteger.valueOf(20_000_000_000L),
			BigInteger.valueOf(200_000),
			HTLC_ADDRESS,
			data,
			BigInteger.ZERO
		).getTransactionHash();
	}

	private TransactionReceipt waitForEthereumReceipt(
		final String transactionHash
	) throws Exception {
		return new PollingTransactionReceiptProcessor(
			ethProvider, 1000, 60).waitForTransactionReceipt(
				transactionHash);
	}

	private static byte[] sha256(
		final byte[] data
		) throws NoSuchAlgorithmException {
			return MessageDigest.getInstance("SHA-256").digest(data);
		}
	}
